pkg/controller/machine-set-boot-image/vsphere_helpers.go

```diff
@@ -0,0 +1,376 @@
+	}
+	datastore, err := finder.Datastore(ctx, providerSpec.Workspace.Datastore)
+	if err != nil {
+		return fmt.Errorf("failed to find datastore: %w", err)
+	}
+
+	// Create mapping between OVF and the network object
+	// found by Name
+	networkMappings := []types.OvfNetworkMapping{{
+		Name:    ovfEnvelope.Network.Networks[0].Name,
+		Network: networkRef.Reference(),
+	}}
+
+	// This is a very minimal spec for importing an OVF.
+	cisp := types.OvfCreateImportSpecParams{
+		EntityName:       name,
+		NetworkMapping:   networkMappings,
+		DiskProvisioning: string(types.OvfCreateImportSpecParamsDiskProvisioningTypeThin),
+	}
+
+	m := ovf.NewManager(client.Client)
+	spec, err := m.CreateImportSpec(ctx,
+		string(ovfDescriptor),
+		resourcePool.Reference(),
+		datastore.Reference(),
+		cisp)
+
+	if err != nil {
+		return fmt.Errorf("failed to create import spec: %w", err)
+	}
+	if spec.Error != nil {
+		return errors.New(spec.Error[0].LocalizedMessage)
+	}
+
+	hostSystem, err := findAvailableHostSystems(ctx, clusterHostSystems, networkRef, datastore)
+	if err != nil {
+		return fmt.Errorf("failed to find available host system: %w", err)
+	}
+
+	lease, err := resourcePool.ImportVApp(ctx, spec.ImportSpec, folder, hostSystem)
+
+	if err != nil {
+		return fmt.Errorf("failed to import vapp: %w", err)
+	}
+
+	info, err := lease.Wait(ctx, spec.FileItem)
+	if err != nil {
+		return fmt.Errorf("failed to lease wait: %w", err)
+	}
+
+	u := lease.StartUpdater(ctx, info)
+	defer u.Done()
+
+	for _, i := range info.Items {
+		// upload the vmdk to which ever host that was first
+		// available with the required network and datastore.
+		err = upload(ctx, archive, lease, i)
+		if err != nil {
+			return fmt.Errorf("failed to upload: %w", err)
+		}
+	}
+
+	err = lease.Complete(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to lease complete: %w", err)
+	}
+
+	vm := object.NewVirtualMachine(client.Client, info.Entity)
+	if vm == nil {
+		return fmt.Errorf("error VirtualMachine not found, managed object id: %s", info.Entity.Value)
+	}
+	if secureBoot {
+		bootOptions, err := vm.BootOptions(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get boot options: %w", err)
+		}
+		bootOptions.EfiSecureBootEnabled = ptr.To(false)
+
+		err = vm.SetBootOptions(ctx, bootOptions)
+		if err != nil {
+			return fmt.Errorf("failed to set boot options: %w", err)
+		}
+	}
+
+	err = vm.MarkAsTemplate(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to mark vm as template: %w", err)
+	}
+
+	err = attachTag(ctx, tagManager, vm.Reference().Value, infraID)
+	if err != nil {
+		return fmt.Errorf("failed to attach tag: %w", err)
+	}
+
+	for _, tagID := range providerSpec.TagIDs {
+		err = attachTag(ctx, tagManager, vm.Reference().Value, tagID)
+		if err != nil {
+			return fmt.Errorf("failed to attach tag: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func getClientsFromServerURL(ctx context.Context, server string) (*govmomi.Client, *tags.Manager, error) {
```

Comment:
Where is the password to vCenter here? And afaik the vcenter url does not have the username in the URI

Reply:
In the installer we started to move to capv's implementation of retrieving govmomi clients because it provides all the session handling out of the box

https://github.com/openshift/installer/blob/main/pkg/asset/installconfig/vsphere/metadata.go

https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/pkg/session/session.go#L116